### ISSUE TITLE
Add ability to sort keys.

### ISSUE THREAD

**Issue by tech4him1** [jakwings/tomlify-j0.4#4]
It would be nice if there were the ability to sort keys with a custom function (like `tomlify(content, replacer, sortKeys)` or something). Is there a possibility that this could be added?

**Comment by iology** [comment]
Could make it something like `tomlify(content, {sort: (path, compare)=>... })`. @adamvoss would it be a problem for typescript?

If you are eager to do this, just copy the single file `tomlify.js` and edit the code around line 329.

**Comment by tech4him1** [comment]
@iology Yes, that would work great. I'm not familar with TypeScript, so I'll wait and let someone else do it. Thanks!

**Comment by iology** [comment]
The module is just vanilla JavaScript. The TypeScript wrapper is from contributors (i.e. I didn't test it myself. ;-)

**Comment by tech4him1** [comment]
@iology Oh, then I'll have to check it out. What interface are you wanting the sort function to have? for example, `js-yaml`'s sort function works like `Array.prototype.sort()`:
```js
/* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort */
(a, b) => {
  if (a is less than b by some ordering criterion) {
    return -1;
  }
  if (a is greater than b by the ordering criterion) {
    return 1;
  }
  // a must be equal to b
  return 0;
}
```

**Comment by iology** [comment]
Right, just that. The new code is on branch v3: https://github.com/jakwings/tomlify-j0.4/tree/v3

**Comment by tech4him1** [comment]
Awesome, I'll check it out! Thanks for working on this so quickly.

**Comment by tech4him1** [comment]
Unrelated, but are you planning on updating this package to keep it in sync with the TOML spec when there are updates?

**Comment by iology** [comment]
> Unrelated, but are you planning on updating this package to keep it in sync with the TOML spec when there are updates?

Yes, only if TOML get big improvements. However its design process seems stagnated, although there are some small changes to 0.4.0.

(i.e. once the tag "wtf" for this repo is removed)

**Comment by adamvoss** [comment]
Sorry about neglecting this, I use GitHub's notification area to keep track of things to work on, but accidentally enabled showing images in GitHub e-mails on my new phone which kept this feature from working (would clear because I read the email).  I'll try to look at whatever I was tagged for this evening.

**Comment by adamvoss** [comment]
Passing a function (or an object containing a function) won't be a problem for TypeScript.

Ultimately, you can express any JavaScript interface in the TypeScript definition, it just becomes a matter of how precise your typings can be and how difficult the typings become to write.  The main areas I have encountered some difficulty in the past is modules/exports and in things that use  `arguments` within the function to make the call signature very flexible.

___

Reading into this further looks like most of this has been implemented, so now there is the task of writing a typings file for the v3 branch?

**Comment by iology** [comment]
> so now there is the task of writing a typings file for the v3 branch?

Yes, I think so.

**Comment by tech4him1** [comment]
@iology Do you have the 3.0.0 branch released to NPM at all (as a alpha/beta version)?

**Comment by iology** [comment]
Published: 3.0.0-alpha.0

**Comment by tech4him1** [comment]
Thank you! What is your plan for v3, will it be something that will take a while, or will it be stable relatively soon? Just wondering, it's no big deal.

**Comment by iology** [comment]
Personally I just use a carbon copy for my private projects, so the current latest release may last longer. :man_shrugging: 

@adamvoss Are you still using 2.x? If you don't mind I will release 3.0.0 earlier without the TypeScript definition file.

**Comment by tech4him1** [comment]
@iology I'm fine with using the alpha version, I don't have any problem with doing that. I just wanted to see if it was something that was going to have huge breaking changes before release or something like that.

**Comment by iology** [comment]
Yeah, it is quiet stable and nobody else have suggestions about 2.x. I guess most people just use this library to dump arbitrary data.

3.x just make it easier to add more options. So people can just stick to 2.x if they don't care for better readability (with sorted keys) of TOML files.

**Comment by iology** [comment]
I have just published v3.0.0 (without TypeScript support). 😉
